### PR TITLE
Add `Artifact::link_with` API to optionally override ELF relocation type

### PR DESCRIPTION
**Pull request by pchickey**

### CODE REVIEW

File: src/elf.rs [hunk 1]
```diff
@@ -437,23 +437,35 @@ impl<'a> Elf<'a> {
...
+                // NB: this now forces _all_ function references, whether local or not, through the PLT
+                // although we're not in the worst company here: https://github.com/ocaml/ocaml/pull/1330
+                Decl::Function {..} => (reloc::R_X86_64_PLT32, -4),
+                Decl::Data {..} => (reloc::R_X86_64_PC32, 0),
+                Decl::FunctionImport => (reloc::R_X86_64_PLT32, -4),
+                Decl::DataImport => (reloc::R_X86_64_GOTPCREL, -4),
+            }
+        };
+
+        let sym_idx = match *to_type {
```
Comment (m4b) [inline src/elf.rs]: I don’t like matching twice but this is fine for now

File: src/artifact.rs [hunk 2]
```diff
@@ -278,12 +286,22 @@ impl Artifact {
     /// **NB**: If either `link.from` or `link.to` is undeclared, then this will return an error.
     /// If `link.from` is an import you previously declared, this will also return an error.
     pub fn link<'a>(&mut self, link: Link<'a>) -> Result<(), Error> {
+        self.link_aux(link, None)
+    }
+    /// A variant of `link` with a RelocOverride provided. Has all of the same invariants as
```
Comment (m4b) [inline src/artifact.rs]: If we keep (probably will) this we need to add note that this is kind of really unsafe unless you know exactly what you’re doing ;)

**Comment by m4b** [comment]
Actually why did you pin the goblin version ? It should be the latest, #205e0ffcc iirc. 

If you are having trouble building try : `cargo update -p goblin` or deleting your lock file

**Comment by pchickey** [comment]
The goblin thing was an issue a few days ago, maybe because the scroll version was wrong. Not sure but its resolved now. Pushed a rebase on your master that doesn't touch Cargo.toml. Sorry about that

**Pull request merged**